docs/concepts/stateful-stream-processing.md

```diff
@@ -0,0 +1,412 @@
+input records while they are storing a snapshot of their state in the *state
+backend*. This *synchronous* state snapshot introduces a delay every time a
+snapshot is taken.
+
+It is possible to let an operator continue processing while it stores its state
+snapshot, effectively letting the state snapshots happen *asynchronously* in
+the background. To do that, the operator must be able to produce a state object
+that should be stored in a way such that further modifications to the operator
+state do not affect that state object. For example, *copy-on-write* data
+structures, such as are used in RocksDB, have this behavior.
+
+After receiving the checkpoint barriers on its inputs, the operator starts the
+asynchronous snapshot copying of its state. It immediately emits the barrier to
+its outputs and continues with the regular stream processing. Once the
+background copy process has completed, it acknowledges the checkpoint to the
+checkpoint coordinator (the JobManager). The checkpoint is now only complete
+after all sinks have received the barriers and all stateful operators have
+acknowledged their completed backup (which may be after the barriers reach the
+sinks).
+
+See [State Backends]({{ site.baseurl }}{% link ops/state/state_backends.md %})
+for details on the state snapshots.
+
+### Recovery
+
+Recovery under this mechanism is straightforward: Upon a failure, Flink selects
+the latest completed checkpoint *k*. The system then re-deploys the entire
+distributed dataflow, and gives each operator the state that was snapshotted as
+part of checkpoint *k*. The sources are set to start reading the stream from
+position <i>S<sub>k</sub></i>. For example in Apache Kafka, that means telling
+the consumer to start fetching from offset <i>S<sub>k</sub></i>.
+
+If state was snapshotted incrementally, the operators start with the state of
+the latest full snapshot and then apply a series of incremental snapshot
+updates to that state.
+
+See [Restart Strategies]({{ site.baseurl }}{% link dev/task_failure_recovery.md
+%}#restart-strategies) for more information.
+
+### State Backends
+
+`TODO: expand this section`
+
+The exact data structures in which the key/values indexes are stored depends on
+the chosen [state backend]({{ site.baseurl }}{% link
+ops/state/state_backends.md %}). One state backend stores data in an in-memory
+hash map, another state backend uses [RocksDB](http://rocksdb.org) as the
+key/value store.  In addition to defining the data structure that holds the
+state, the state backends also implement the logic to take a point-in-time
+snapshot of the key/value state and store that snapshot as part of a
+checkpoint.
+
+<img src="{{ site.baseurl }}/fig/checkpoints.svg" alt="checkpoints and snapshots" class="offset" width="60%" />
+
+{% top %}
+
+### Savepoints
+
+`TODO: expand this section`
+
+Programs written in the Data Stream API can resume execution from a
```

Comment:
all streaming programs, right?

Reply:
hence all programs that use checkpointing

Reply:
yes 👌